text/0000-ledger-api.md

```diff
@@ -0,0 +1,380 @@
+
+To allow for functional differences between different but very similar operations the stubAPI has chosen an approach to have a 'broad-and-flat' approach to APIs. For example there are query APIs per type of collection, `getPrivateDataByRange` and `getStateByRange`, by apply modifiers to the function name the function is extended. The length of the resulting list of APIs is larger and should additional function become available will increase. A long list of APIs increases the difficulty of finding the API needed, especially without autocomplete-style tool support. By formalizing the APIs in the proposed structure makes an overall more sustainable model to support future changes.  
+
+Where required, invalid operations can be detected and a 'fail-fast' approach used. Developers would therefore be fully aware of the limitations during basic testing.
+
+Future RFCs will focus on the persistence of the business data; many of the same helper classes used today are based around the idea of serializing data, and move the knowledge of key creation into the business objects themselves. (c.f. https://jira.hyperledger.org/browse/FABCN-92)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A complete solution using Fabric can be represented as follows:
+![](0000-ledger-api-ConceptualArchitecture.png)
+
+A typical sequence is for an application to connect to a peer (the 'gateway' into the network channel), and obtain a reference to a smart contract. The Network-shared transaction logic in the Smart Contract is then be invoked by name with optional data arguments and return values. The data arguments can include transient data if needed. The logic in the Smart Contract will typical reference and update the shared transaction data, subject of course to exactly how the code is written.
+
+The 'Ledger' is, therefore, a primary abstraction in a blockchain system, including in Hyperledger Fabric; this abstraction allows the concepts of states, world state, collections, transactions, blocks etc. to be tied together and formalised. 
+
+Looking at the objects that make up this conceptual architecture, a full set of objects would have the following relationships. 
+
+![](0000-ledger-api-relationships.png)
+
+Note that this RFC is concerned with the ledger-api and the ledger,collection and state; the 'white' boxes will be subsequent RFCs.
+
+## Ledger
+
+A Ledger's current value is held in a set of collections. A single ledger is available to any given deployed contract, though other ledgers may be available within the Network Channel and can be accessed via other contracts
+
+Of the collections that are reference by the ledger, there a single collection in the ledger that is public (the 'WorldState') and held by every organization in the Network Channel. A ledger may have 1 or more private collections that are available only in some organizations. Starting at Fabric v2.0 there will be an implicit private collection per organization.
```

Comment:
Not sure whether we can call both the world state and private state as a collection. As far as I know, the term collection has been used only for private data. By collection, I always thought that the actual data gets into a collection and the hash of the data gets into the block. Whatever data present in the block is called world state is what I understood. Please correct me if I have misunderstood.

Reply:
Your understanding of Fabric is correct - and this API is not changing anything about the concept. 
Today for example if you look at putState() and putPrivateState() there are different APIs, but in terms of the gRPC flows to the peer - it's the same low-level flow.. the gRPC level-putstate, takes a string, which is the empty string for the worldstate, and a name for private data.